sharding/interfaces.go

```diff
@@ -1,20 +1,62 @@
 package sharding
 
-// Service defines items that can be registered into a sharding node's serviceFuncs.
-//
-// life-cycle management is delegated to the sharding node. The service is allowed to
-// initialize itself upon creation, but no goroutines should be spun up outside of the
-// Start method.
-type Service interface {
-	// Start is called after all services have been constructed and the networking
-	// layer was also initialized to spawn any goroutines required by the service.
+import (
+	"fmt"
+	"reflect"
+)
+
+// Node defines a a sharding-enabled Ethereum instance that provides
+// full control and shared access of necessary components and services
+// for a sharded Ethereum blockchain.
+type Node interface {
 	Start() error
+	Close() error
+	Register(constructor ServiceConstructor) error
+}
+
+// ShardP2P defines an interface for a peer-to-peer service in a
+// sharded Ethereum blockchain.
+type ShardP2P interface{}
 
-	// Stop terminates all goroutines belonging to the service, blocking until they
-	// are all terminated.
+// TXPool defines an interface for a transaction pool service that handles
+// incoming shard transactions in the network.
+type TXPool interface{}
+
+// Actor refers to either a notary, proposer, or observer in the sharding spec.
+type Actor interface {
+	Start() error
 	Stop() error
 }
 
-// ServiceConstructor defines the callback passed in when registering a service
-// to a sharding node.
-type ServiceConstructor func() (Service, error)
+// ServiceContext is a collection of service independent options inherited from
+// the protocol stack, that is passed to all constructors to be optionally used;
+// as well as utility methods to operate on the service environment.
+type ServiceContext struct {
+	Services map[reflect.Type]Service // Index of the already constructed services
+}
+
+// ServiceConstructor is the function signature of the constructors needed to be
+// registered for service instantiation.
+type ServiceConstructor func(ctx *ServiceContext) (Service, error)
+
+// Service is an individual protocol that can be registered into a node. Having a sharding
+// node maintain a service registry allows for easy, shared-dependencies. For example,
+// a proposer service might depend on a p2p server, a txpool, an smc client, etc.
+type Service interface {
+	// Start is called after all services have been constructed to
+	// spawn any goroutines required by the service.
```

Comment:
Please add a note that this must be a non-blocking method or other services will not be started.